### PR TITLE
By default, disable the context's PTY on Windows platform

### LINKED ISSUE #339: When `exit_code` or `run` is invoked, the task does not complete (`fwrite()` error)
Castor 0.14.0 (Phar)
PHP 8.3.4
PowerShell 7.4.1 on Windows 11

Everytime I try to run a task that includes a call to `run()` or `exit_code()`, the task does not complete. I have to either `Ctrl+C` or press `Enter` to exit the task, and then an error is displayed:

```
In AbstractPipes.php line 139:

  Notice: fwrite(): Write of 2 bytes failed with errno=22 Invalid argument
```

The command is OK, because I can see its output right after this error message.
A sample task triggering this error:

```php
#[AsTask]
function phpVersion(): void {
  run("php --version"); // Using a string or an array as argument does not change anything.
}
```

This error does not occur if I use Symfony Process directly (i.e. `(new Process(["php", "--version"]))->run()`).

### PR DESCRIPTION
**Pull request by cedx**
Fixes #339

### CODE REVIEW

**Comment by lyrixx** [comment]
Hello @TheoD02 ; What do you think of this one?

**Comment by TheoD02** [comment]
> Hello @TheoD02 ; What do you think of this one?

Hey,

I've never used castor on Windows but only via WSL

I can test this evening to reproduce the case if necessary, I can not confirm at this time

**Comment by TheoD02** [comment]
> Fixes #339

How did you get castor to work on Windows?

It doesn't work for me, with version `8.3.4`, `8.2.17` with the following modules enabled : 
```
[PHP Modules]
bcmath
bz2
calendar
Core
ctype
curl
date
dom
exif
FFI
fileinfo
filter
ftp
gd
gettext
gmp
hash
iconv
imap
intl
json
ldap
libxml
mbstring
mysqli
mysqlnd
odbc
openssl
pcre
PDO
pgsql
Phar
random
readline
Reflection
session
shmop
SimpleXML
soap
sockets
sodium
SPL
sqlite3
standard
tidy
tokenizer
xml
xmlreader
xmlwriter
xsl
zip
zlib
```

I tried on my machine and via a sandbox, same result…

I tried with phar and vendor composer.

As soon as I launch castor, it freezes over and over again :/
image

**Comment by cedx** [comment]
It freezes only when it doesn't find the "castor.php" file. So you must launch it from your project root.
(I'm on vacation, with only my phone, so I can't post more details for now).

**Comment by TheoD02** [comment]
Ok thanks for the answer!

I'll test again, making sure I have a castor.php file at the root ^^

I'll see about debugging the castor.php detection part that's blocking, happy vacations! 😁

**Comment by TheoD02** [comment]
> Hello @TheoD02 ; What do you think of this one?

Hello,

I have just retested, it is correct for me. It can be merged :)

By default, when a command is run, nothing happens until the command is exited via `CTRL`+`C`. `PTY` defined to false for Windows fix the problem 😄

File: src/Context.php [hunk 1]
```diff
@@ -28,6 +29,7 @@ public function __construct(
         // Do not use this argument, it is only used internally by the application
         public readonly string $name = '',
     ) {
+        $this->pty = $pty ?? (\PHP_OS_FAMILY != 'Windows');
```
Comment (lyrixx) [inline src/Context.php:32]:
```suggestion
        $this->pty = $pty ?? (\PHP_OS_FAMILY !== 'Windows');
```

**Review by pyrech — APPROVED** [review]
Thanks for the fix :slightly_smiling_face:

**Pull request closed without merging**